[Bug #7080] Improve documentation for custom blocks in gridBlock

shashank0470 opened the pull request:
### Type
📝 Bug Fix / Documentation Update

### Description
This PR addresses Issue #7080 by improving the inline documentation in `Blocks.js` for configuring custom blocks in the `gridBlock`.

It clarifies that in addition to adding a block to `allowedBlocks`, developers must also extend `gridBlock.blocksConfig` with the block's configuration to ensure proper rendering within the grid layout.

### What Changed
- Added clear and detailed instructions in the comments of `Blocks.js`
- Explained why both `allowedBlocks` and `blocksConfig` need to be updated
- Provided a complete example showing the correct configuration pattern

### Why It Matters
Many developers were only adding blocks to `allowedBlocks`, which resulted in blocks not functioning properly within grid layouts. This fix prevents that confusion and improves developer onboarding.

### Related Issue
Closes #7080

shashank0470 committed:
docs: clarify how to configure custom blocks in gridBlock Added detailed explanation on why both llowedBlocks and locksConfig updates are necessary to enable custom blocks inside gridBlock. This change helps developers avoid misconfigurations when extending grid blocks in add-ons.

shashank0470 committed:
docs(changelog): add changelog entry for #7080

stevepiercy commented:
This is a code comment, not documentation. It should fit in with existing documentation at https://6.docs.plone.org/volto/blocks/core/grid.html#block-configuration.

shashank0470 commented:
Hello @stevepiercy 
Thanks for the feedback. Just to clarify — could you please point me to the exact documentation file or location where this update should go? I want to make sure I contribute it in the right place

stevepiercy commented:
@shashank0470 you can go about it two ways: through the `plone/volto` repository as you've done here, or through the `plone/documentation` repository. For the former method, see  [Documentation](https://6.docs.plone.org/volto/contributing/documentation.html), and the latter see [Set up, build, and check the quality of documentation](https://6.docs.plone.org/contributing/documentation/setup-build.html).

The file to edit is https://github.com/plone/volto/blob/main/docs/source/blocks/core/grid.md

shashank0470 committed:
docs(grid.md): clarify custom block config in gridBlock (#7080) Improved documentation to show how to correctly configure custom blocks inside gridBlock, including both 'allowedBlocks' and 'blocksConfig' usage. Added news entry.

shashank0470 commented:
Thanks, @stevepiercy. I've moved the explanation from the code comments into grid.md as suggested.
Also added a news entry under packages/volto/news/7080.documentation.
Let me know if anything else needs changes!

stevepiercy commented on docs/source/blocks/core/grid.md:
````diff
@@ -76,13 +76,30 @@ The grid block has the following configuration settings.
     You could add different variations, `schemaEnhancers`, and so on.
     You could remove them as well, but only for blocks inside the grid block.
 
+    **Adding custom blocks to the grid:**
+    
+    When adding a custom block (like `countUpBlock`) to the grid, you need **both** configurations:
+
+    ```js
+    // 1. Add the block to allowedBlocks
+    config.blocks.blocksConfig.gridBlock.allowedBlocks.push('countUpBlock');
+    
+    // 2. Add the block configuration to grid's blocksConfig
+    config.blocks.blocksConfig.gridBlock.blocksConfig = {
+      ...config.blocks.blocksConfig.gridBlock.blocksConfig,
+      countUpBlock: config.blocks.blocksConfig.countUpBlock,
+    };
+    ```
+
+    **Customizing existing blocks in the grid:**
+
````
This addition is for two of the grid block configuration settings, so it should be a new subsection and not within the `blocksConfig` configuration setting alone.

````suggestion

### Add custom blocks to a grid block

When adding a custom block—for example, `countUpBlock`—to the grid block, you need to use both the `allowedBlocks` and `blocksConfig` configuration settings.

First, add the block to `allowedBlocks`.

```js
config.blocks.blocksConfig.gridBlock.allowedBlocks.push('countUpBlock');
```

Then add the block configuration to the grid block's `blocksConfig`.

```js
config.blocks.blocksConfig.gridBlock.blocksConfig = {
  ...config.blocks.blocksConfig.gridBlock.blocksConfig,
  countUpBlock: config.blocks.blocksConfig.countUpBlock,
};
```



### Customize existing blocks in a grid block

````

I'm not sure what's going on here between this heading and the code that follows it. I'm also not sure whether it's needed, given the previous examples, but maybe it is useful.

In any case, keep the original version under the `blocksConfig` setting and indented, then duplicate and edit the code block so that it's under the new subheading, is dedented, and has both the `allowedBlocks` and `blocksConfig`. It also needs an introductory sentence to describe what's going on. Does that make sense?

stevepiercy commented on packages/volto/news/7080.bugfix:
```diff
@@ -0,0 +1 @@
+Clarified gridBlock documentation to explain the need for both allowedBlocks and blocksConfig configuration. @shashank0470
```
Also rename this file to `packages/volto/news/7080.documentation`.

```suggestion
Clarified the need for both `allowedBlocks` and `blocksConfig` configuration settings to add a block to a grid block. @shashank0470
```

shashank0470 committed:
docs: clarify grid block custom block configuration requirements

- Move custom block configuration guide from code comments to grid.md documentation
- Add new 'Add custom blocks to a grid block' subsection with clear examples
- Clarify that both allowedBlocks and blocksConfig are required for custom blocks
- Improve note formatting and language for better clarity
- Follow Plone documentation style with one sentence per line
- Rename news file from 7080.bugfix to 7080.documentation
Fixes configuration confusion where developers only set allowedBlocks without adding block configuration to grid's blocksConfig.

shashank0470 committed:
chore: remove redundant news file

shashank0470 committed:
docs(grid): refine grid block documentation and remove redundant changelog file

- Revised the documentation in  grid.md based on reviewer feedback
- Reformatted examples for clarity and consistency
- Removed the redundant  news/7080.bugfix  file
- Ensured all necessary details are included only in  7080.documentation

shashank0470 committed:
fix: apply Prettier and ESLint fixes to Blocks.jsx

stevepiercy commented on docs/source/blocks/core/grid.md:
````diff
@@ -100,6 +100,57 @@ The grid block has the following configuration settings.
+};
+```
+
+**Customizing existing blocks in the grid:**
````
```suggestion

### Customize existing blocks in a grid block
```

stevepiercy commented on docs/source/blocks/core/grid.md:
```diff
@@ -100,6 +100,57 @@ The grid block has the following configuration settings.
+
+**Customizing existing blocks in the grid:**
+
+You can also customize how existing blocks behave within the grid:
```
```suggestion
The following example shows how to customize the behavior of existing blocks within the grid.
```

stevepiercy commented on docs/source/blocks/core/grid.md:
````diff
@@ -100,6 +100,57 @@ The grid block has the following configuration settings.
     };
     ```
 
+### Add custom blocks to a grid block
````
```suggestion

(blocks-core-grid-add-custom)=
### Add custom blocks to a grid block
```

stevepiercy commented on packages/volto/src/config/Blocks.jsx:
```diff
@@ -520,14 +520,29 @@ const blocksConfig = {
+//    };
+//
+// The second step is required because the grid block uses its own blocksConfig
+// to render inner blocks, and simply adding to allowedBlocks is not sufficient.
```
I don't know if this comment block is useful here. It adds a maintenance burden for documentation when there is already narrative documentation and keeping the same content in two places. I think it's better to provide a link to the specific section of the documentation with details, and remove all the comments. That link will be:

https://6.docs.plone.org/volto/blocks/core/grid.html#blocks-core-grid-add-custom

So maybe replace it all with this:

```
// IMPORTANT: When adding custom blocks to the grid block in add-ons, see the documentation for requirements at
// https://6.docs.plone.org/volto/blocks/core/grid.html#blocks-core-grid-add-custom
```

shashank0470 replied:
Hi @stevepiercy,

I’ve made all the requested changes:
- Removed the verbose comment block and replaced it with a concise reference link to the official documentation for adding custom blocks to `gridBlock`.

All tests are now passing as well.  
Please review and merge when convenient.

Thanks again for your patience!

shashank0470 replied:
Hi @stevepiercy ,

I've implemented the changes and manually tested the functionality — everything seems to be working as expected, and I haven’t encountered any crashes or visible issues.

At the same time, I’m still getting familiar with the overall code structure and testing setup, so there's quite a lot for me to understand and align with right now. I’ve been trying to figure out the correct way to write tests in this context, but I’m still facing some challenges.

Would it be okay if I get some guidance on how best to proceed with the test coverage? I want to make sure I do it the right way and meet the project’s standards.

Thanks a lot!

shashank0470 committed:
Refactor: Clean up gridBlock comment with concise link to documentation

stevepiercy commented on docs/source/blocks/core/grid.md:
````diff
@@ -100,6 +100,61 @@ The grid block has the following configuration settings.
+    ],
+  },
+};
+```
````
OK, but did you test this code sample to verify that it actually works?

stevepiercy reviewed (approved):
I approve, but we need technical verification. Can we get an assist? @ksuess raised this issue, so maybe she can help?

shashank0470 commented:
Hi @ksuess 

Thank you for the feedback. I understand that this issue is more complex and not a current priority for the team.

I’m still in the process of understanding the codebase and appreciate this learning experience. If it makes more sense to close this PR and focus on more suitable issues aligned with the current Volto roadmap, I’m happy to do that.

Please let me know how you'd prefer I proceed — I’m here to contribute wherever I can be most helpful.
